### PR TITLE
[MRG] Fix SBT duplicates bug

### LINKED ISSUE #1171: duplication of signatures seen in large SBT databases
start with https://github.com/dib-lab/sourmash/issues/849#issuecomment-666895041 and go from there :)

### PR DESCRIPTION
**Pull request by ctb**
For Zip SBTs, signatures with identical md5sums were being saved correctly as different signatures, but the updated signature filenames resulting from duplicate md5sums were not being communicated back to the tree properly. This meant that the *names* of a number of signatures (but not their content) were being lost, triggering the behavior seen over in https://github.com/dib-lab/sourmash/issues/1511#issuecomment-846146110 as well as a number of apparent duplicate signatures per #1171.

Fixes #1171.

TODO:
- [x] add basic tests that trigger the problem!
- [x] add tests where we see what happens if the SBT already exists ;)

### CODE REVIEW

**Comment by ctb** [comment]
The tests are failing because of the need to support `--append` for zip SBTs; the revised code is saving the index as `.sbt.json_0`. I am thinking of adding a new function to SBT storage that is `save_exact`, which will replace the contents of the file if needed without generating a new filename.

File: tests/test-data/duplicate-sigs/README.md [hunk 1]
```diff
@@ -0,0 +1,2 @@
+This directory contains two signatures with different metadata but the same
```
Comment (ctb) [inline tests/test-data/duplicate-sigs/README.md]:
```suggestion
This directory contains four signatures with different metadata but the same
```

**Comment by ctb** [comment]
OK, @luizirber, I think this is ready for a look.

File: src/sourmash/sbt.py [hunk 2]
```diff
@@ -286,6 +287,8 @@ def add_node(self, node):
             c1 = self.children(p.pos)[0]
             self._leaves[c1.pos] = node 
             node.update(n)
+        else:
+            assert 0
```
Comment (luizirber) [inline src/sourmash/sbt.py:292]: maybe be more explicit why this assertion is here, it will be hard to figure out in the future =]

Comment (ctb) [inline src/sourmash/sbt.py:292]:
```suggestion
            # this branch should never be reached!
            assert 0
```

File: src/sourmash/sbt_storage.py [hunk 3]
```diff
@@ -19,6 +20,9 @@ def save(self, path, content):
     def load(self, path):
         pass
 
+    def save_exact(self, path, content):
```
Comment (luizirber) [inline src/sourmash/sbt_storage.py]: what do you think about changing the signature of `save` to
```
def save(self, path, content, *, overwrite=False):
```
and avoid `save_exact`?

Comment (ctb) [reply src/sourmash/sbt_storage.py]: sure! I was thinking about that, too, and went with this for no particular reason - will adjust.

**Comment by ctb** [comment]
@luizirber there's one failing test under python 3.9 on linux - ipfsstorage. I'm unclear on whether it's something I broke, tho.

File: src/sourmash/sbt_storage.py [hunk 4]
```diff
@@ -12,7 +12,7 @@
 class Storage(ABC):
 
     @abc.abstractmethod
-    def save(self, path, content):
+    def save(self, path, content, overwrite=False):
```
Comment (luizirber) [inline src/sourmash/sbt_storage.py]:
```suggestion
    def save(self, path, content, *, overwrite=False):
```

**Comment by luizirber** [comment]
> @luizirber there's one failing test under python 3.9 on linux - ipfsstorage. I'm unclear on whether it's something I broke, tho.

It's something that was broken: `list_sbts` is not implemented for IPFSStorage (but wasn't tested/checked/needed before, because IPFS-backed SBTs usually are a single JSON anyway)
https://github.com/dib-lab/sourmash/pull/1568/checks?check_run_id=2748963142#step:12:135

(actually implementing it is interesting because it leads to using an IPFS address as we use .zip files nowadays, hmm...)

**Comment by ctb** [comment]
> > @luizirber there's one failing test under python 3.9 on linux - ipfsstorage. I'm unclear on whether it's something I broke, tho.
> 
> It's something that was broken: `list_sbts` is not implemented for IPFSStorage (but wasn't tested/checked/needed before, because IPFS-backed SBTs usually are a single JSON anyway)
> https://github.com/dib-lab/sourmash/pull/1568/checks?check_run_id=2748963142#step:12:135
> 
> (actually implementing it is interesting because it leads to using an IPFS address as we use .zip files nowadays, hmm...)

OK. Any thoughts on what to do about it, or why it's acting up now? I can dig into it later, just wondering if you know off the top of your head.

**Comment by ctb** [comment]
yay w00t! fixed!

**Comment by ctb** [comment]
(ready for review and merge @luizirber)

File: src/sourmash/sbt_storage.py [hunk 5]
```diff
@@ -113,50 +119,58 @@ def __init__(self, path):
...
                     n += 1
+            except KeyError:
+                return testpath, True
 
-        zf.writestr(newpath, content)
-        return newpath
+        assert 0 # should never get here!
 
-    def save(self, path, content):
+    def save(self, path, content, overwrite=False):
```
Comment (luizirber) [inline src/sourmash/sbt_storage.py]:
```suggestion
    def save(self, path, content, *, overwrite=False):
```

File: src/sourmash/sbt_storage.py [hunk 6]
```diff
@@ -268,7 +282,7 @@ def __init__(self, pin_on_add=True, **kwargs):
         self.pin_on_add = pin_on_add
         self.api = ipfshttpclient.connect(**self.ipfs_args)
 
-    def save(self, path, content):
+    def save(self, path, content, overwrite=False):
```
Comment (luizirber) [inline src/sourmash/sbt_storage.py]:
```suggestion
    def save(self, path, content, *, overwrite=False):
```

File: src/sourmash/sbt_storage.py [hunk 7]
```diff
@@ -306,7 +320,7 @@ def __init__(self, **kwargs):
         self.redis_args = kwargs
         self.conn = redis.Redis(**self.redis_args)
 
-    def save(self, path, content):
+    def save(self, path, content, overwrite=False):
```
Comment (luizirber) [inline src/sourmash/sbt_storage.py]:
```suggestion
    def save(self, path, content, *, overwrite=False):
```

**Comment by luizirber** [comment]
> OK. Any thoughts on what to do about it, or why it's acting up now? I can dig into it later, just wondering if you know off the top of your head.

Trying to think a bit conceptually about what is happening when saving/loading:

- a `location` is always required. It can be a JSON file (with the SBT description, for `FSStorage` and `IPFSStorage`) or a Zip file containing the SBT JSON (for `ZipStorage`).
- `storage` is not required, but can be passed to **replace** the storage defined in the `location`. For example, we can use an SBT with a `FSStorage` in `location`, but actually use a `ZipStorage` to load/save data.

Part of the confusion in using this API is that we mostly depend on the extension of the `location` to decide what to do. I'll try to come up with a table later to map behaviors...

**Review by luizirber — APPROVED** [review]

**Comment by ctb** [comment]
> > OK. Any thoughts on what to do about it, or why it's acting up now? I can dig into it later, just wondering if you know off the top of your head.
> 
> Trying to think a bit conceptually about what is happening when saving/loading:
> 
>     * a `location` is always required. It can be a JSON file (with the SBT description, for `FSStorage` and `IPFSStorage`) or a Zip file containing the SBT JSON (for `ZipStorage`).
> 
>     * `storage` is not required, but can be passed to **replace** the storage defined in the `location`. For example, we can use an SBT with a `FSStorage` in `location`, but actually use a `ZipStorage` to load/save data.
> 
> 
> Part of the confusion in using this API is that we mostly depend on the extension of the `location` to decide what to do. I'll try to come up with a table later to map behaviors...

Right - and I have a much better understanding of the code's behavior now 😄 , which is nice!

**Pull request merged**